### ISSUE TITLE
Confusing display when content served from Chrome's cache

### ISSUE THREAD

**Issue by thed000d** [janodvarko/harviewer#61]
When I look at this request that was served directly from Chrome's cache (not from the server with a 304 header) I see its size listed 658 B. Hovering over the size shows `0 (0 bytes) - compressed, 658 B (658 bytes) - uncompressed`. The 658 B are included in the summary row as the total size and are *not* included in the "From Cache" calculation. 

Here is the HAR entry:
```
      {
        "startedDateTime": "2016-05-16T18:25:05.667Z",
        "time": 13.887000037357211,
        "request": {
          "method": "GET",
          "url": "https://URL/assets/jquery.delegate-16afa9bbc5fdd25927818d4e3fbb09e7.js",
          "httpVersion": "unknown",
          "headers": [
            {
              "name": "Accept",
              "value": "*/*"
            },
            {
              "name": "Referer",
              "value": "https://URL/"
            },
            {
              "name": "User-Agent",
              "value": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36"
            }
          ],
          "queryString": [],
          "cookies": [],
          "headersSize": -1,
          "bodySize": 0
        },
        "response": {
          "status": 200,
          "statusText": "OK",
          "httpVersion": "unknown",
          "headers": [
            {
              "name": "Date",
              "value": "Mon, 16 May 2016 18:24:19 GMT"
            },
            {
              "name": "Content-Encoding",
              "value": "gzip"
            },
            {
              "name": "Server",
              "value": "Apache"
            },
            {
              "name": "Vary",
              "value": "Accept-Encoding"
            },
            {
              "name": "Content-Type",
              "value": "application/javascript"
            },
            {
              "name": "Cache-Control",
              "value": "max-age=31536000"
            },
            {
              "name": "Accept-Ranges",
              "value": "bytes"
            },
            {
              "name": "Content-Length",
              "value": "321"
            },
            {
              "name": "Expires",
              "value": "Tue, 16 May 2017 18:24:19 GMT"
            }
          ],
          "cookies": [],
          "content": {
            "size": 658,
            "mimeType": "application/javascript",
            "text": "JAVASCRIPT"
          },
          "redirectURL": "",
          "headersSize": -1,
          "bodySize": 0,
          "_transferSize": 0
        },
        "cache": {},
        "timings": {
          "blocked": 6.09800004167482,
          "dns": -1,
          "connect": -1,
          "send": 0,
          "wait": 4.39499999629338,
          "receive": 3.3939999993890115,
          "ssl": -1
        },
        "pageref": "page_7"
      }
```

With a non-cached request, HAR Viewer shows the amount of data actually transferred (compressed size) for the request and in the summary statistics. To be consistent, I'd expect HAR Viewer to show the amount of data actually transferred (0 bytes) or some indication that the request came from the cache.

**Comment by gitgrimbo** [comment]
Hi, I think agree with your point about showing 0 bytes for a response taken from the browser cache.

But are you sure that the HAR entry you show is a cached entry?  The `timings` values not being `0` or `-1`, and the `response.content.text` having a value make me a bit suspicious (especially if this HAR came from Chrome dev tools, which in my experience uses `0` and `-1` in its `timings` for a cached entry).

@janodvarko, is there a foolproof way to determine if a [HAR entry](https://github.com/gitgrimbo/harviewer/blob/2027922c9e1a60ce40b50050ff8a2a7045aae0e2/webapp/scripts/preview/harSchema.js#L89) represents a cached entry or not?  Or are HAR entries **always** expected to be the result of a genuine network request?

Assuming we have to determine if the entry is from cache or not, for Chrome it seems we can work this out from the `timings` data, which *seems* to only contain `0` or `-1` for the times, e.g.:

        "timings": {
          "blocked": -1,
          "dns": -1,
          "connect": -1,
          "send": 0,
          "wait": 0,
          "receive": 0,
          "ssl": -1
        },

From memory, Firefox (using [netExport](https://getfirebug.com/wiki/index.php/Firebug_Extensions#NetExport)) didn't export the cached entries at all, and so this issue #61 wouldn't be an issue at all.

Could this be an issue with Chrome?  Should Chrome **not** be exporting HAR entries that were satisfied by browser cache?

**Comment by thed000d** [comment]
I'm sure it is a cached entry. Here's what I get when I look at the actual Chrome Developer Tools for this requests: 

![screen shot 2016-05-25 at 10 40 27 am](https://cloud.githubusercontent.com/assets/1156111/15548363/41660a34-2265-11e6-8983-576c4fa06855.png)

**Comment by gitgrimbo** [comment]
Extra investigation as to how to determine if a HAR entry came from cache.

**TL;DR** - it doesn't look like we can determine if a HAR entry is from the browser cache just by looking at `entry.time` and `entry.timings`.  So the question still stands, **is there a foolproof way to determine if a HAR entry represents a cached entry or not?**

The [`entry.cache`](http://www.softwareishard.com/blog/har-12-spec/#cache) field is empty for every entry in the HAR captured below.  **Is this correct/expected?**

-----

**First load of harviewer** - The following image shows a clean-cache load of harviewer.

![harviewer-61-first-load](https://cloud.githubusercontent.com/assets/1020541/15551724/36d1d996-22af-11e6-9eff-768c764718bc.png)

**Second load of harviewer** - The following image shows a second load of harviewer.  There are 4 resources that come from Chrome's cache.

![harviewer-61-second-load](https://cloud.githubusercontent.com/assets/1020541/15551730/3db89c4a-22af-11e6-8f56-1a2180b07305.png)

**HAR from the second load** - The following file is the HAR captured from the second load (using Chrome dev tools - *Copy all as HAR*).

[harviewer-61-second-load.har.txt](https://github.com/janodvarko/harviewer/files/282649/harviewer-61-second-load.har.txt)

**Showing the HAR from the second load** - The following image shows the second load's HAR file open in harviewer.  We can see that of the 4 resources that came from Chrome's cache, 3 of them had a `0ms` total time, and one had `2ms` total time.  So this result proves my theory that Chrome always uses `0ms` time for cached entries is false.

![harviewer-61-showing-har-for-second-load](https://cloud.githubusercontent.com/assets/1020541/15551736/427dd146-22af-11e6-8ccf-85bf34837f96.png)

**Comment by thed000d** [comment]
What about looking at the `response.bodySize` and `response.headerSize`? A `bodySize` of 0 seems to indicate it came from cache (especially if the `content.size` is non-zero). A `headerSize` of -1 and/or a `send` timing of 0 seem to indicate that there wasn't even a request made to a server.

**Comment by gitgrimbo** [comment]
These 3 entries from https://gist.github.com/gitgrimbo/2173aaab1a12efe9233323078d6e7c97#file-harviewer-61-second-load-har-js don't seem to give a clear picture unless I'm mistaken.

http://www.softwareishard.com/har/viewer/css/harViewer.css  (from browser cache, has `response.headersSize > 0`)

    time: 0
    request.headersSize: 758
    response.status: 200
    response.content.size: 23640
    response.headersSize: 349,
    response.bodySize: 0,
    response._transferSize: 0

http://www.google-analytics.com/ga.js  (from browser cache, has `response.headersSize == -1`)

    time: 2.0140000001447333
    request.headersSize: -1
    response.status: 200
    response.content.size: 4308
    response.headersSize: -1,
    response.bodySize: 0
    response._transferSize: 0

http://www.softwareishard.com/har/viewer/scripts/harViewer.js  (HTTP 304, `response.bodySize == 0`. We can rule out this entry as from browser cache because `response.status == 304`, unless browsers can cache HTTP 304 responses, so the HTTP 304 entries themselves came from the browser cache??)

    time: 53.93800000001647
    request.headersSize: 842
    response.status: 304
    response.content.size: 313020
    response.headersSize: 241,
    response.bodySize: 0
    response._transferSize: 241

**Comment by gitgrimbo** [comment]
Background.  Chrome DevTools `_transferSize` custom field in the wild:

- https://codereview.chromium.org/825243002/
- https://github.com/cyrus-and/chrome-har-capturer/commit/67982eb663043f3c2745e53efbb5c73f14850e6c

**Comment by gitgrimbo** [comment]
@thed000d - can you check http://gitgrimbo.github.io/harviewer/issue-61/ with your HARs to see if this is fixed for you?

A resource fulfilled by the browser cache or by a HTTP 304 should:
    
- Appear as a grey row in the RequestList.
- Show the text "From Cache" in the EntrySizeInfoTip.
- Contribute to the "From Cache" segment of the cache pie chart.
- Contribute to the "From Cache" part of the RequestList summary.

**Comment by thed000d** [comment]
Yes, I checked this and it looks like a huge improvement! Nice work. I'd feel good about closing this issue. 

There is still one inconsistency. When a request isn't cached harviewer utilizes the compressed size of the data transferred. When a request is cached harviewer uses the uncompressed size (and there is no way to know the compressed size). So when a page is first loaded it looks like the amount of data transferred is _less_ than when it is loaded for a second time (with items in the cache). You can see this in your `issue 61 chrome51*` tests by summing the `cached` and `downloaded` components of the `cache` element and noting that they are larger when using the cache. 

Again, this is a huge improvement so up to you how (or if) you want to address the display of how much data is transferred.

**Comment by gitgrimbo** [comment]
Cool, nice to hear.

I've also felt uneasy with the different numbers coming from Chrome, Firefox and IE11 HAR export with respect to resource sizes, so I think I understand your point.

As you have said, the Chrome, Firefox and IE11 HARs included with the https://github.com/gitgrimbo/harviewer/tree/issue-61/tests/hars/issue-61 branch, give the following numbers for the `ga.js` resource:

    Use case            content.size  body.size  content.compression  _transferSize
    Firefox no cache           43082      16022
    Firefox Ctrl+R (304)       43082      16022
    Firefox no request         43082          0

    Chrome no cache            43082      16022                27060          16435
    Chrome Ctrl+R (304)        43082          0                                 117
    Chrome no request          43082          0                                   0

    IE11 no cache              43082      43082
    IE11 Ctrl+R (304)              0          0
    IE11 no request [1]            0          0

[1] Actually there does seem to be a request made (and the response is 304).  This could be something weird to do with IE11's cache being at O/S level and not a 'local' browser cache?

Based on these numbers, I'm not sure there's a consistent view of compressed content size and uncompressed content size when a resource comes from cache.  I think the only consistent way is to always show 0 bytes for the compressed and uncompressed size for a cached resource (which is accurate as no bytes were transferred) - otherwise it means different behaviour of HAR Viewer depending on which browser captured the HAR (*but is this an issue? not sure as I'm not familiar with the original requirements here*).

So in the case of Chrome and Firefox when there is no request made, we no longer have the compressed resource size in the HAR (for that specific request), and so the totals/summary/pie charts will *necessarily* be different, and be larger, than the original no-cache request and the 304 request.  Should the non-IE browsers at least be consistent here?

Also, from what I can see, no browser uses the [cache object from the HAR spec](http://www.softwareishard.com/blog/har-12-spec/#cache).

I'd like to merge the changes, but I'll await your response first to see if there are further changes to be made.
